### DISCUSSION TITLE
长草流程的暂停与缓停功能

### DISCUSSION THREAD

**Discussion by YipKei** [MaaAssistantArknights/MaaAssistantArknights#2091] [Ideas]
一键长草目前终止执行流只有两种方式，要么是等全部任务执行完成，要么是强行结束当前执行，就好像刹车只能直接拉手刹一样，感觉二者之间需要一些折中的选项，缓停和暂停是其中的两个设想。

-----

# 缓停

所谓缓停，是相对于当前的  `停止` 操作的一种操作，点击缓停后，执行完当前进行中的最小行为组之后结束操作流。

最小行为组指的是一个在玩家实际操作中执行事件的最小单元，比如刷图一次，其行为组会在点击掉结算页面之后结束。

点击缓停之后，缓停过程结束后才会释放被锁定的项目。

缓停功能可以确保MAA的操作被正常记录，同时，缓停的执行模式也是后续暂停的前置需求。

# 暂停

暂停功能和缓停类似，但在结束当前操作之后，会额外进行一次截图，记录当前状态与执行中的任务列表和一些需要记录的信息（比如肉鸽的配置单等）。

玩家使用暂停，并等待暂停完成后，可以操作游戏进行一些其他的操作。

操作完成后，只要恢复到暂停时的界面，MAA比对前后的状态（当然，忽略理智信息这种可能会随时间发生微弱变化的场景），如果确认前后一致，那么会接续暂停前的任务继续执行，反之则抛出异常。

玩家如果无法恢复之前暂停的状态，则可以选择结束当前流程，此时起到的效果相当于缓停。

暂停功能可以让玩家在连续挂肉鸽的时候临时中断进程，去执行一些简单的操作（比如收个菜，或者打两把芯片/红票本），然后回来继续上一把肉鸽，而不是要等到当前对局战斗失败，或者强行结束当前对局。

# 其他设想

如果实现了缓停和暂停功能，那么肉鸽可以增加一个断点续打的选项，如果之前肉鸽是通过缓停模式保存了进度，那么MAA会自动读取进度接续进行上一局的游戏，反之则按现在的流程重开，相当于挂肉鸽更灵活了。

**Comment by zzyyyl** [comment]
其实按照现在的逻辑肉鸽完全可以继续的，不知道为啥MAA一定要从头打

**Reply by zzyyyl** [reply]
肉鸽最早的时候就是可以继续的来着

**Reply by MistEO** [reply]
改了啊，现在地图界面能进了
